Improve the premium plugins purchase flow

mmtr opened the pull request:
Fixes https://github.com/Automattic/wp-calypso/issues/65499

#### Proposed Changes

This PR improves the overall flow for purchasing premium plugins by applying these changes:
- Avoid making requests to fetch the Atomic transfer status if there is already an ongoing request
- Hide the "Back to plugins" link if the plugin has not been installed yet, effectively allowing us to partially revert changes from https://github.com/Automattic/wp-calypso/pull/65293
- Change the link of the "Manage plugins" button for Limited Atomic sites since they cannot access `wp-admin/plugins.php`
- Simplify the loading screen in order to account for cases where the site was already Atomic before purchasing a plugin

#### Testing Instructions

- Use the Calypso live link below
- Switch to Simple site with a Free plan
- Go to Plugins
- Purchase a WooCommerce extension
- Make sure the loading screen goes through 4 steps and it doesn't display the "Back to plugins" button
- In the "Thank you" page, make sure the "Manage plugins" button redirects you to `/plugins/:plugin/:site`
- Make sure that the "WooCommerce" menu items have been updated
- Purchase any other WooCommerce extension
- Make sure the loading screen has now 1 single step since the site is already Atomic
- Switch to Simple site with a Pro plan
- Install a free plugin
- Make sure you only see one loading screen with 3 steps before being redirected to the "Thank you" page
- In the "Thank you" page, make sure the "Manage plugins" button redirects you to `wp-admin/plugins.php`
- Switch to a self-hosted Jetpack site
- Install a free plugin
- In the "Thank you" page, make sure the "Manage plugins" button is enabled and redirects you to `wp-admin/plugins.php`

#### Pre-merge Checklist

Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in [Simple](P9HQHe-k8-p2-p2), [Atomic](P9HQHe-jW-p2-p2), and [self-hosted Jetpack sites](PCYsg-g6b-p2-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Are we memoizing when appropriate (for expensive computations)? More info in [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md) and [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors)
- [ ] Have we sent any new strings [for translation](PCYsg-1vr-p2-p2) ASAP?

mmtr committed:
Improve the premium plugins purchase flow

mmtr committed:
Update menu after plugin has been installed

mmtr committed:
Revert isFetchingAutomatedTransferStatus selector changes

mmtr committed:
Hide progress bar when coming from the free plugin install flow since that has its own loading screen

mmtr committed:
Do no fetch transfer status on self-hosted Jetpack sites

cpapazoglou commented on client/my-sites/checkout/checkout-thank-you/marketplace/marketplace-thank-you.tsx:
```diff
@@ -178,9 +175,11 @@ const MarketplaceThankYou = ( { productSlug }: { productSlug: string } ) => {
 	const fallbackSetupUrl = wpComPluginData?.setup_url && siteAdminUrl + wpComPluginData?.setup_url;
+	const managePluginsUrl = hasManagePluginsFeature
+		? `${ siteAdminUrl }plugins.php`
+		: `/plugins/manage/${ siteSlug } `;
```
Let's avoid using `plugins/manage`. We 've been using `/plugins.php` in case we didn't have a specific WP Admin URL (action_link). Can we replace it with `/plugins/pluginSlug`?

mmtr replied:
`wp-admin/plugins.php` is only available for sites with the `MANAGE_PLUGINS` feature, so we need to redirect to something else for the rest of sites (e.g. Atomic sites with a Starter plan). See 837-gh-Automattic/wpcomsh

cpapazoglou replied:
Yep, got that part. That's why I suggest to redirect to Calypso `/plugins/pluginSlug/siteId`

mmtr replied:
Oh sorry, I read your suggestion too fast and thought you were suggesting to revert the changes 🙂 Yes, that could work!

mmtr committed:
Use plugin screen as destination of Manage Plugins button on Limited Atomic sites (props @cpapazoglou)

obenland commented:
Tested with a Starter-Plan site purchasing a premium plugin:

* It looks like the ProgressBar and the All Set screen are shown simultaneously with a scroll bar <img width="1510" alt="Screen Shot 2022-07-18 at 9 53 31 AM" src="https://user-images.githubusercontent.com/1398304/179563629-97b67683-b585-40e0-800a-52ad3122b37a.png">
* The removal of the `sawPluginOnce` logic re-introduced the ProgressBar flash at the end of the transfer process, when there's a plugins request with an empty response. However I was only able to reproduce it on my first try, subsequent plugin installations (even on other sites) did not show that behavior.
* Installing another plugin after the site has gone Atomic, the Progress bar starts at Step 2 of 2, installing the plugin <img width="1510" alt="Screen Shot 2022-07-18 at 10 08 05 AM" src="https://user-images.githubusercontent.com/1398304/179565222-6145304f-989b-4f3c-ac4d-9039e5e6d522.png">
* Cutting the progress steps to 2 results in a pretty long delay in step 1 as we're waiting for the transfer to complete (20 sec in my test). We had done 4 steps to give more of an impression of progress. 

https://user-images.githubusercontent.com/1398304/179566467-01228cf7-06f9-4a32-acd0-4d1c2f2c30ba.mov

mmtr committed:
Hide the "All ready to go" screen while showing the progress bar

mmtr committed:
Fix progress bar flashing

mmtr commented:
> It looks like the ProgressBar and the All Set screen are shown simultaneously with a scroll bar

Nice catch! Fixed in 676073338114fca37251190a1bb20806e78fcdbc

> The removal of the sawPluginOnce logic re-introduced the ProgressBar flash at the end of the transfer process

I added back an alternative and simpler logic in 1badb6f3f7265a23f41754fd8149a14a76eb0511 that should avoid this

> Installing another plugin after the site has gone Atomic, the Progress bar starts at Step 2 of 2, installing the plugin

Yeah, this also happens currently in production, with the difference that it starts at Step 4 of 4. This happens because the site is already Atomic, so all the steps covering the Atomic transfer are completed immediately. The problem is that we don't know whether we show this screen for an already transferred or if the site is being transferred.

One option would be to re-arrange the steps once we know that info, but that will cause a flash in the rendering process.

> We had done 4 steps to give more of an impression of progress.

Yeah, the problem is that those 4 steps cause more confusion when the site is already transferred.

mmtr committed:
Keep using 4 steps during Atomic transfers

mmtr commented:
> One option would be to re-arrange the steps once we know that info, but that will cause a flash in the rendering process.

Turns out we already know when the component gets rendered for the first time if the site is already Atomic, so in 194ff39dc185f105da6ff1994b22e064e7b6ff7e I adapted the steps for those cases.

obenland commented on client/my-sites/checkout/checkout-thank-you/marketplace/marketplace-thank-you.tsx:
```diff
@@ -109,58 +123,66 @@ const MarketplaceThankYou = ( { productSlug }: { productSlug: string } ) => {
-		if ( transferStatus === AtomicTransferActive ) {
+		// Sites already transferred to Atomic or self-hosted Jetpack sites only show one step.
+		if ( isJetpack ) {
+			setCurrentStep( 0 );
```
On first transfer of a site, I think it might be this conditional that causes a flash of "Step 1 of 4" at the end of the plugin installation process, when the site has fully transferred?

https://user-images.githubusercontent.com/1398304/179801476-72aa1ff6-2c53-40ac-b33d-267069dd7dfe.mov

mmtr replied:
Another good catch! I think I fixed it in 2dd89b6.

obenland reviewed (commented):
It's almost perfect!

mmtr committed:
Do not change current step if site is already Atomic

obenland reviewed (approved):
Thank you!
